File: src/main/java/com/example/PowerUp.java
import java.awt.*;
import java.awt.geom.Ellipse2D;
import java.util.ArrayList;
import java.util.Random;
import javax.swing.Timer;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;

public class PowerUp extends Control {
    // Constants
    public static final int DIAMETER = 20;
    private static final double SPEED = 3.5; // Speed of the power-up
    private Random random = new Random();
    private int timeFreeze = random.nextInt(25) + 1;
    private static final int FRAME_WIDTH = 800;
    private static final int FRAME_HEIGHT = 600;

    // Constructor
    public PowerUp(int x, int y) {
        super(x, y, 0, 0, DIAMETER, DIAMETER); // Call to the parent constructor with parameters
    }

    // Draw the power-up
    public void draw(Graphics g) {
        g.setColor(new Color(0, 206, 209)); // teal
        g.fillOval(x, y, DIAMETER, DIAMETER);
    }

    // Move the power-up downwards
    public void move() {
        y += SPEED; // Move the power-up down by increasing its y coordinate
    
        // If the power-up has moved off the bottom of the screen, reset its position to the top
        if (y > FRAME_HEIGHT) {
            y = 0;
            x = (int) (Math.random() * FRAME_WIDTH); // Generate a new random x coordinate
        }
    }
    public void moveDown(int bottomBound, int gameWidth) {
        y += SPEED;
        if (y > bottomBound) {
            y = 0;
            x = random.nextInt(gameWidth - DIAMETER); // Generate a new x position
        }
    }

    public static PowerUp checkSpawn(int score, boolean gameJustStarted) {
        // Spawn a power-up only if the player's score is a multiple of 50 and the game has not just started
        if (score % 75 == 0 && score != 0 && !gameJustStarted) {
            Random random = new Random();
            int x = random.nextInt(Background.getFrameWidth());
            int y = random.nextInt(Background.getFrameHeight());
            return new PowerUp(x, y);
        }
        return null;
    }

    public boolean isCollidingWithPlayer(Player player) {
        if (getBounds().intersects(player.getBounds())) {
            player.stopMoving();

            // Move the player to the top of the JFrame
            player.setY(0);

            // Slow down the player's fall
            player.slowFall();

            Timer timer = new Timer(5000, new ActionListener() {
                @Override
                public void actionPerformed(ActionEvent e) {
                    player.stopMoving();
                }
            });
            timer.setRepeats(false);
            timer.start();

            return true;
        }
        return false;
    }

    public void collideWithPaddles(ArrayList<Paddle> paddles) {
        for (Paddle paddle : paddles) {
            if (getBounds().intersects(paddle.getBounds())) {
                paddle.speedUp();
                Timer timer = new Timer(5000, new ActionListener() {
                    @Override
                    public void actionPerformed(ActionEvent e) {
                        paddle.resetSpeed();
                    }
                });
                timer.setRepeats(false);
                timer.start();
            }
        }
    }

    // Get the bounding circle of the PowerUp
    public Rectangle getBounds() {
        return new Ellipse2D.Double(x, y, DIAMETER, DIAMETER).getBounds();
    }

    // Getters and setters
    public int getTimeFreeze() {
        return this.timeFreeze;
    }

    public void setTimeFreeze(int timeFreeze) {
        this.timeFreeze = timeFreeze;
    }

    public int getY() {
        return this.y;
    }

    public int getX() {
        return this.x;
    }
}

File: projects/MusicGame/Music-Game-main/src/PowerUp.js
class PowerUp {
    // Constants
    static DIAMETER = 20;
    static SPEED = 3.5; // Speed of the power-up
    static FRAME_WIDTH = 800;
    static FRAME_HEIGHT = 600;

    constructor(x, y) {
        this.x = x;
        this.y = y;
        this.timeFreeze = Math.floor(Math.random() * 25) + 1;
    }

    // Draw the power-up
    draw(ctx) {
        ctx.fillStyle = 'rgb(0, 206, 209)'; // teal
        ctx.beginPath();
        ctx.arc(this.x, this.y, PowerUp.DIAMETER / 2, 0, Math.PI * 2);
        ctx.fill();
    }

    // Move the power-up downwards
    move() {
        this.y += PowerUp.SPEED; // Move the power-up down by increasing its y coordinate

        // If the power-up has moved off the bottom of the screen, reset its position to the top
        if (this.y > PowerUp.FRAME_HEIGHT) {
            this.y = 0;
            this.x = Math.random() * PowerUp.FRAME_WIDTH; // Generate a new random x coordinate
        }
    }

    moveDown(bottomBound, gameWidth) {
        this.y += PowerUp.SPEED;
        if (this.y > bottomBound) {
            this.y = 0;
            this.x = Math.floor(Math.random() * (gameWidth - PowerUp.DIAMETER)); // Generate a new x position
        }
    }

    static checkSpawn(score, gameJustStarted) {
        // Spawn a power-up only if the player's score is a multiple of 75 and the game has not just started
        if (score % 75 === 0 && score !== 0 && !gameJustStarted) {
            const x = Math.floor(Math.random() * PowerUp.FRAME_WIDTH);
            const y = Math.floor(Math.random() * PowerUp.FRAME_HEIGHT);
            return new PowerUp(x, y);
        }
        return null;
    }

    isCollidingWithPlayer(player) {
        if (this.getBounds().intersects(player.getBounds())) {
            player.stopMoving();

            // Move the player to the top of the canvas
            player.setY(0);

            // Slow down the player's fall
            player.slowFall();

            setTimeout(() => {
                player.stopMoving();
            }, 5000);

            return true;
        }
        return false;
    }

    collideWithPaddles(paddles) {
        paddles.forEach(paddle => {
            if (this.getBounds().intersects(paddle.getBounds())) {
                paddle.speedUp();
                setTimeout(() => {
                    paddle.resetSpeed();
                }, 5000);
            }
        });
    }

    // Get the bounding circle of the PowerUp
    getBounds() {
        return {
            x: this.x,
            y: this.y,
            width: PowerUp.DIAMETER,
            height: PowerUp.DIAMETER,
            intersects: function (rect) {
                return !(rect.x > this.x + this.width ||
                         rect.x + rect.width < this.x ||
                         rect.y > this.y + this.height ||
                         rect.y + rect.height < this.y);
            }
        };
    }

    // Getters and setters
    getTimeFreeze() {
        return this.timeFreeze;
    }

    setTimeFreeze(timeFreeze) {
        this.timeFreeze = timeFreeze;
    }

    getY() {
        return this.y;
    }

    getX() {
        return this.x;
    }
}